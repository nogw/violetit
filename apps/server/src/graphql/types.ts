import { ParameterizedContext } from 'koa';
import { Maybe } from '@violetit/types';

import { IUserDocument } from '../modules/user/UserModel';
import { DataLoaders } from '../modules/loader/loaderRegister';

export type GraphQLContext = {
  ctx: ParameterizedContext;
  user?: Maybe<IUserDocument>;
  dataloaders: DataLoaders;
};
