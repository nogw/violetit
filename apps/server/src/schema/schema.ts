import { GraphQLSchema } from 'graphql';

import { MutationType } from './MutationType';
import { QueryType } from './QueryType';

export const schema = new GraphQLSchema({
  mutation: MutationType,
  query: QueryType,
});
