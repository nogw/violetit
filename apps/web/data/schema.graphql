"""
Root of all queries
"""
type Query {
  """
  Fetches an object given its ID
  """
  node(
    """
    The ID of an object
    """
    id: ID!
  ): Node

  """
  Fetches objects given their IDs
  """
  nodes(
    """
    The IDs of objects
    """
    ids: [ID!]!
  ): [Node]!

  """
  Logged user
  """
  me: User

  """
  Connection to all communities
  """
  communities(after: String, first: Int, before: String, last: Int, filters: CommunityFilters): CommunityConnection!
  community(id: String!): Community

  """
  Connection to all posts
  """
  posts(after: String, first: Int, before: String, last: Int, filters: PostFilter): PostConnection!
}

"""
An object with an ID
"""
interface Node {
  """
  The id of the object.
  """
  id: ID!
}

type User implements Node {
  """
  The ID of an object
  """
  id: ID!
  username: String!
  email: String!
  communities(after: String, first: Int, before: String, last: Int): CommunityConnection!
}

"""
A connection to a list of items.
"""
type CommunityConnection implements Connection {
  """
  Number of items in this connection.
  """
  count: Int

  """
  A count of the total number of objects in this connection, ignoring pagination.
  This allows a client to fetch the first five objects by passing "5" as the
  argument to "first", then fetch the total count so it could display "5 of 83",
  for example.
  """
  totalCount: Int

  """
  Offset from start.
  """
  startCursorOffset: Int!

  """
  Offset till end.
  """
  endCursorOffset: Int!

  """
  Information to aid in pagination.
  """
  pageInfo: PageInfo!

  """
  A list of edges.
  """
  edges: [CommunityEdge]!
}

"""
A connection to a list of items.
"""
interface Connection {
  """
  Number of items in this connection.
  """
  count: Int

  """
  A count of the total number of objects in this connection, ignoring pagination.
  This allows a client to fetch the first five objects by passing "5" as the
  argument to "first", then fetch the total count so it could display "5 of 83",
  for example.
  """
  totalCount: Int

  """
  Offset from start.
  """
  startCursorOffset: Int!

  """
  Offset till end.
  """
  endCursorOffset: Int!

  """
  Information to aid in pagination.
  """
  pageInfo: PageInfo!
}

"""
Information about pagination in a connection.
"""
type PageInfo {
  """
  When paginating forwards, are there more items?
  """
  hasNextPage: Boolean!

  """
  When paginating backwards, are there more items?
  """
  hasPreviousPage: Boolean!

  """
  When paginating backwards, the cursor to continue.
  """
  startCursor: String

  """
  When paginating forwards, the cursor to continue.
  """
  endCursor: String
}

"""
An edge in a connection.
"""
type CommunityEdge {
  """
  The item at the end of the edge.
  """
  node: Community

  """
  A cursor for use in pagination.
  """
  cursor: String!
}

type Community implements Node {
  """
  The ID of an object
  """
  id: ID!

  """
  mongoose _id
  """
  _id: String!
  createdAt: String
  updatedAt: String
  name: String!
  title: String!
  admin: ID!
  mods: [ID]!
  tags(after: String, first: Int, before: String, last: Int): TagConnection!
  members(after: String, first: Int, before: String, last: Int): UserConnection!
  joined: Boolean!
}

"""
A connection to a list of items.
"""
type TagConnection implements Connection {
  """
  Number of items in this connection.
  """
  count: Int

  """
  A count of the total number of objects in this connection, ignoring pagination.
  This allows a client to fetch the first five objects by passing "5" as the
  argument to "first", then fetch the total count so it could display "5 of 83",
  for example.
  """
  totalCount: Int

  """
  Offset from start.
  """
  startCursorOffset: Int!

  """
  Offset till end.
  """
  endCursorOffset: Int!

  """
  Information to aid in pagination.
  """
  pageInfo: PageInfo!

  """
  A list of edges.
  """
  edges: [TagEdge]!
}

"""
An edge in a connection.
"""
type TagEdge {
  """
  The item at the end of the edge.
  """
  node: Tag

  """
  A cursor for use in pagination.
  """
  cursor: String!
}

type Tag implements Node {
  """
  The ID of an object
  """
  id: ID!
  label: String!
  color: String!
  createdBy: User
  community: Community

  """
  mongoose _id
  """
  _id: String!
  createdAt: String
  updatedAt: String
}

"""
A connection to a list of items.
"""
type UserConnection implements Connection {
  """
  Number of items in this connection.
  """
  count: Int

  """
  A count of the total number of objects in this connection, ignoring pagination.
  This allows a client to fetch the first five objects by passing "5" as the
  argument to "first", then fetch the total count so it could display "5 of 83",
  for example.
  """
  totalCount: Int

  """
  Offset from start.
  """
  startCursorOffset: Int!

  """
  Offset till end.
  """
  endCursorOffset: Int!

  """
  Information to aid in pagination.
  """
  pageInfo: PageInfo!

  """
  A list of edges.
  """
  edges: [UserEdge]!
}

"""
An edge in a connection.
"""
type UserEdge {
  """
  The item at the end of the edge.
  """
  node: User

  """
  A cursor for use in pagination.
  """
  cursor: String!
}

"""
Used to filter communities
"""
input CommunityFilters {
  OR: [CommunityFilters]
  AND: [CommunityFilters]

  """
  Order reviews by DateOrderingInputType.
  """
  orderBy: [DateOrdering!]

  """
  Filter by search.
  """
  search: String
}

"""
Input to order by createdAt or updatedAt.
"""
input DateOrdering {
  """
  Field used to sort, e.g. CREATED_AT.
  """
  field: DateEnumType!

  """
  Direction used to sort, e.g. ASC.
  """
  direction: DirectionEnum!
}

enum DateEnumType {
  """
  creation date
  """
  CREATED_AT

  """
  document update date
  """
  UPDATED_AT
}

enum DirectionEnum {
  ASC
  DESC
}

"""
A connection to a list of items.
"""
type PostConnection implements Connection {
  """
  Number of items in this connection.
  """
  count: Int

  """
  A count of the total number of objects in this connection, ignoring pagination.
  This allows a client to fetch the first five objects by passing "5" as the
  argument to "first", then fetch the total count so it could display "5 of 83",
  for example.
  """
  totalCount: Int

  """
  Offset from start.
  """
  startCursorOffset: Int!

  """
  Offset till end.
  """
  endCursorOffset: Int!

  """
  Information to aid in pagination.
  """
  pageInfo: PageInfo!

  """
  A list of edges.
  """
  edges: [PostEdge]!
}

"""
An edge in a connection.
"""
type PostEdge {
  """
  The item at the end of the edge.
  """
  node: Post

  """
  A cursor for use in pagination.
  """
  cursor: String!
}

type Post implements Node {
  """
  The ID of an object
  """
  id: ID!
  createdAt: String
  updatedAt: String
  title: String!
  content: String!
  author: User
  community: Community
  votesCount: Int!
  meHasVoted: Vote
}

type Vote implements Node {
  """
  The ID of an object
  """
  id: ID!
  createdAt: String
  updatedAt: String
  type: String!
  post: ID!
  user: ID!
}

"""
Used to filter posts
"""
input PostFilter {
  OR: [PostFilter]
  AND: [PostFilter]

  """
  Order reviews by DateOrderingInputType.
  """
  orderBy: [DateOrdering!]

  """
  Filter by community.
  """
  community: ID
}

"""
Root of all mutations
"""
type Mutation {
  userRegister(input: UserRegisterInput!): UserRegisterPayload
  userLogin(input: UserLoginInput!): UserLoginPayload
  communityCreate(input: CommunityCreateInput!): CommunityCreatePayload
  communityJoin(input: CommunityJoinInput!): CommunityJoinPayload
  communityExit(input: CommunityExitInput!): CommunityExitPayload
  communityExitAsAdmin(input: CommunityExitAsAdminInput!): CommunityExitAsAdminPayload
  postCreate(input: PostCreateInput!): PostCreatePayload
  voteCreate(input: VoteCreateInput!): VoteCreatePayload
  tagCreateMutation(input: TagCreateInput!): TagCreatePayload
  tagDeleteMutation(input: TagDeleteInput!): TagDeletePayload
}

type UserRegisterPayload {
  token: String
  me: User

  """
  Default success field resolver.
  """
  success: String
  error: FieldError
  clientMutationId: String
}

type FieldError {
  field: String
  message: String
}

input UserRegisterInput {
  username: String!
  email: String!
  password: String!
  clientMutationId: String
}

type UserLoginPayload {
  token: String
  me: User

  """
  Default success field resolver.
  """
  success: String
  error: FieldError
  clientMutationId: String
}

input UserLoginInput {
  email: String!
  password: String!
  clientMutationId: String
}

type CommunityCreatePayload {
  communityEdge: CommunityEdge
  error: FieldError

  """
  Default success field resolver.
  """
  success: String
  clientMutationId: String
}

input CommunityCreateInput {
  name: String!
  title: String!
  clientMutationId: String
}

type CommunityJoinPayload {
  communityEdge: CommunityEdge
  error: FieldError

  """
  Default success field resolver.
  """
  success: String
  clientMutationId: String
}

input CommunityJoinInput {
  communityId: String!
  clientMutationId: String
}

type CommunityExitPayload {
  communityEdge: CommunityEdge
  error: FieldError

  """
  Default success field resolver.
  """
  success: String
  clientMutationId: String
}

input CommunityExitInput {
  communityId: String!
  clientMutationId: String
}

type CommunityExitAsAdminPayload {
  communityEdge: CommunityEdge
  error: FieldError

  """
  Default success field resolver.
  """
  success: String
  clientMutationId: String
}

input CommunityExitAsAdminInput {
  communityId: String!
  clientMutationId: String
}

type PostCreatePayload {
  postEdge: PostEdge
  error: FieldError

  """
  Default success field resolver.
  """
  success: String
  clientMutationId: String
}

input PostCreateInput {
  tags: [ID]!
  title: String!
  content: String!
  community: String!
  clientMutationId: String
}

type VoteCreatePayload {
  post: Post
  vote: VoteEdge

  """
  Default success field resolver.
  """
  success: String
  error: FieldError
  clientMutationId: String
}

"""
An edge in a connection.
"""
type VoteEdge {
  """
  The item at the end of the edge
  """
  node: Vote

  """
  A cursor for use in pagination
  """
  cursor: String!
}

input VoteCreateInput {
  postId: String!
  type: VoteType
  clientMutationId: String
}

enum VoteType {
  UPVOTE
  DOWNVOTE
}

type TagCreatePayload {
  tagEdge: TagEdge

  """
  Default success field resolver.
  """
  success: String
  error: FieldError
  clientMutationId: String
}

input TagCreateInput {
  label: String!
  color: String!
  communityId: String!
  clientMutationId: String
}

type TagDeletePayload {
  """
  Default success field resolver.
  """
  success: String
  error: FieldError
  clientMutationId: String
}

input TagDeleteInput {
  tagId: String!
  communityId: String!
  clientMutationId: String
}
