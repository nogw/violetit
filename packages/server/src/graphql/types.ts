import { DataLoaders } from '../modules/loader/loaderRegister';
import { IUser } from '../modules/user/UserModel';
import { Maybe } from '@violetit/types';

export type GraphQLContext = {
  user?: Maybe<IUser>;
  dataloaders: DataLoaders;
};
