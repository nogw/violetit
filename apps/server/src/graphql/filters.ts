import { GraphQLEnumType, GraphQLInputObjectType, GraphQLNonNull } from 'graphql';
import { DirectionEnumType, SortDirection } from '@entria/graphql-mongo-helpers';

type DateSort = 'createdAt' | 'updatedAt';

type DateOrdering = {
  field: DateSort;
  direction: SortDirection;
};

const dateFields = {
  CREATED_AT: {
    value: 'createdAt',
    description: 'creation date',
  },
  UPDATED_AT: {
    value: 'updatedAt',
    description: 'document update date',
  },
};

const DateEnumType = new GraphQLEnumType({
  name: 'DateEnumType',
  values: dateFields,
});

const DateOrderingInputType = new GraphQLInputObjectType({
  name: 'DateOrdering',
  description: 'Input to order by createdAt or updatedAt.',
  fields: () => ({
    field: {
      type: new GraphQLNonNull(DateEnumType),
      description: 'Field used to sort, e.g. CREATED_AT.',
    },
    direction: {
      type: new GraphQLNonNull(DirectionEnumType),
      description: 'Direction used to sort, e.g. ASC.',
    },
  }),
});

export { dateFields, DateSort, DateOrdering, DateEnumType, DateOrderingInputType };
