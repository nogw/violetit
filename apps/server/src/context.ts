import { Maybe } from '@violetit/types';
import { ParameterizedContext } from 'koa';

import { getDataloaders } from './modules/loader/loaderRegister';
import { GraphQLContext } from './graphql/types';
import { IUserDocument } from './modules/user/UserModel';

type ContextVars = {
  ctx: ParameterizedContext;
  user: Maybe<IUserDocument>;
};

export const getContext = async ({ ctx, user }: ContextVars): Promise<GraphQLContext> => {
  const dataloaders = getDataloaders();

  return {
    ctx,
    dataloaders,
    user,
  } as const;
};
