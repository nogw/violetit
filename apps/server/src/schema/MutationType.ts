import { GraphQLObjectType } from 'graphql';

import * as userMutations from '../modules/user/mutations';
import * as communityMutations from '../modules/community/mutations';
import * as postMutations from '../modules/post/mutations';
import * as voteMutations from '../modules/vote/mutations';
import * as tagMutations from '../modules/tag/mutations';

export const MutationType = new GraphQLObjectType({
  name: 'Mutation',
  description: 'Root of all mutations',
  fields: () => ({
    ...communityMutations,
    ...userMutations,
    ...postMutations,
    ...voteMutations,
    ...tagMutations,
  }),
});
