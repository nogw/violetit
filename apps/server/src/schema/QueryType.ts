import { GraphQLNonNull, GraphQLObjectType } from 'graphql';
import { connectionArgs } from '@entria/graphql-mongo-helpers';

import { nodeField, nodesField } from '../modules/node/typeRegister';

import { UserType } from '../modules/user/UserType';
import UserLoader from '../modules/user/UserLoader';

import { PostFiltersInputType } from '../modules/post/PostFilterInputType';
import { PostConnection } from '../modules/post/PostType';
import PostLoader from '../modules/post/PostLoader';

import { CommunityFiltersInputType } from '../modules/community/CommunityFilterInputType';
import { CommunityConnection } from '../modules/community/CommunityType';
import CommunityLoader from '../modules/community/CommunityLoader';

export const QueryType = new GraphQLObjectType({
  name: 'Query',
  description: 'Root of all queries',
  fields: () => ({
    node: nodeField,
    nodes: nodesField,

    me: {
      type: UserType,
      description: 'Logged user',
      resolve: (_root, _args, context) => {
        return UserLoader.load(context, context.user?._id);
      },
    },

    communities: {
      type: new GraphQLNonNull(CommunityConnection.connectionType),
      description: 'Connection to all communities',
      args: {
        ...connectionArgs,
        filters: {
          type: CommunityFiltersInputType,
        },
      },
      resolve: async (_, args, context) => CommunityLoader.loadAll(context, args),
    },

    posts: {
      type: new GraphQLNonNull(PostConnection.connectionType),
      description: 'Connection to all posts',
      args: {
        ...connectionArgs,
        filters: {
          type: PostFiltersInputType,
        },
      },
      resolve: async (_, args, context) => PostLoader.loadAll(context, args),
    },
  }),
});
