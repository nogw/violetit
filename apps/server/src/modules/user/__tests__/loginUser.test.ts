import { graphql } from 'graphql';

import {
  clearDbAndRestartCounters,
  connectWithMongoose,
  disconnectWithMongoose,
  sanitizeTestObject,
} from '../../../../test';
import { createUser } from '../fixtures/createUser';
import { schema } from '../../../schema/schema';
import { getContext } from '../../../context';

beforeAll(connectWithMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectWithMongoose);

describe('UserLoginMutation', () => {
  it('should login with a registered user', async () => {
    const { username, email } = await createUser({
      email: 'nogw@nogw.com',
      password: 'a9218c490c89864790bf',
    });

    const mutation = `
      mutation UserLoginMutation($input: UserLoginInput!) {
        userLogin(input: $input) {
          token
          error {
            message
          }
          me {
            username
          }
        }
      }
    `;

    const rootValue = {};
    const contextValue = getContext();
    const variables = {
      email: email,
      password: 'a9218c490c89864790bf',
    };

    const result = await graphql({
      schema,
      rootValue,
      contextValue,
      source: mutation,
      variableValues: { input: variables },
    });

    expect(result.errors).toBeUndefined();
    expect(result.data.userLogin.error).toBeNull();
    expect(result.data.userLogin.token).toBeDefined();
    expect(result.data.userLogin.me.username).toBe(username);
    expect(sanitizeTestObject(result.data, ['token'])).toMatchSnapshot();
  });

  it("should display error if username isn't exists", async () => {
    const mutation = `
      mutation UserLoginMutation($input: UserLoginInput!) {
        userLogin(input: $input) {
          token
          error {
            message
          }
          me {
            id
            username
          }
        }
      }
    `;

    const rootValue = {};
    const contextValue = getContext();
    const variables = {
      email: 'awesome@fpmail.com',
      password: 'awesomepassword',
    };

    const result = await graphql({
      schema,
      rootValue,
      contextValue,
      source: mutation,
      variableValues: { input: variables },
    });

    expect(result.errors).toBeUndefined();
    expect(result.data.userLogin.token).toBeNull();
    expect(result.data.userLogin.me).toBeNull();
    expect(result.data.userLogin.error.message).toBe('This user was not registered');
    expect(sanitizeTestObject(result.data)).toMatchSnapshot();
  });

  it('should display error if password is incorrect', async () => {
    const { email } = await createUser({
      email: 'violetit@mail.com',
    });

    const mutation = `
      mutation UserLoginMutation($input: UserLoginInput!) {
        userLogin(input: $input) {
          token
          error {
            message
          }
          me {
            id
          }
        }
      }
    `;

    const rootValue = {};
    const contextValue = getContext();
    const variables = {
      email: email,
      password: 'awesomepassword',
    };

    const result = await graphql({
      schema,
      rootValue,
      contextValue,
      source: mutation,
      variableValues: { input: variables },
    });

    expect(result.errors).toBeUndefined();
    expect(result.data.userLogin.token).toBeNull();
    expect(result.data.userLogin.me).toBeNull();
    expect(result.data.userLogin.error.message).toBe('This password is incorrect');
    expect(sanitizeTestObject(result.data)).toMatchSnapshot();
  });
});
