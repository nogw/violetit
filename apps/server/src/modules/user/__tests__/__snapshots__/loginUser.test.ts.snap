// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserLoginMutation should display error if password is incorrect 1`] = `
Object {
  "userLogin": Object {
    "error": Object {
      "message": "This password is incorrect",
    },
    "me": "EMPTY",
    "token": "EMPTY",
  },
}
`;

exports[`UserLoginMutation should display error if username isn't exists 1`] = `
Object {
  "userLogin": Object {
    "error": Object {
      "message": "This user was not registered",
    },
    "me": "EMPTY",
    "token": "EMPTY",
  },
}
`;

exports[`UserLoginMutation should login with a registered user 1`] = `
Object {
  "userLogin": Object {
    "error": "EMPTY",
    "me": Object {
      "username": "user#0",
    },
    "token": "FROZEN-TOKEN",
  },
}
`;
